src/main/java/maksym/fedorenko/bookstore/service/impl/ShoppingCartServiceImpl.java

```diff
@@ -0,0 +1,96 @@
+package maksym.fedorenko.bookstore.service.impl;
+
+import lombok.RequiredArgsConstructor;
+import maksym.fedorenko.bookstore.dto.shoppingcart.CartDto;
+import maksym.fedorenko.bookstore.dto.shoppingcart.CartItemDto;
+import maksym.fedorenko.bookstore.dto.shoppingcart.CreateCartItemRequestDto;
+import maksym.fedorenko.bookstore.dto.shoppingcart.UpdateCartItemRequestDto;
+import maksym.fedorenko.bookstore.exception.EntityNotFoundException;
+import maksym.fedorenko.bookstore.mapper.CartItemMapper;
+import maksym.fedorenko.bookstore.mapper.ShoppingCartMapper;
+import maksym.fedorenko.bookstore.model.Book;
+import maksym.fedorenko.bookstore.model.CartItem;
+import maksym.fedorenko.bookstore.model.ShoppingCart;
+import maksym.fedorenko.bookstore.repository.BookRepository;
+import maksym.fedorenko.bookstore.repository.CartItemRepository;
+import maksym.fedorenko.bookstore.repository.ShoppingCartRepository;
+import maksym.fedorenko.bookstore.service.ShoppingCartService;
+import org.springframework.security.core.Authentication;
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+public class ShoppingCartServiceImpl implements ShoppingCartService {
+    private final BookRepository bookRepository;
+    private final ShoppingCartRepository shoppingCartRepository;
+    private final CartItemRepository cartItemRepository;
+    private final ShoppingCartMapper shoppingCartMapper;
+    private final CartItemMapper cartItemMapper;
+
+    @Override
+    public CartDto getUserCart(Authentication authentication) {
+        String username = authentication.getName();
+        ShoppingCart cart = shoppingCartRepository.findByUserEmailWithCartItems(username)
+                .orElseThrow(() -> getNotFoundCartException(username));
+        return shoppingCartMapper.toDto(cart);
+    }
+
+    @Override
+    public CartItemDto addCartItem(
+            Authentication authentication, CreateCartItemRequestDto requestDto) {
+        ShoppingCart cart = retrieveUserCartFromDb(authentication.getName());
+        CartItem cartItem = cartItemRepository
+                .findByShoppingCartAndBookId(cart, requestDto.bookId())
+                .orElseGet(() -> createCartItemFromRequest(requestDto, cart));
+        if (cartItem.getId() != null) {
+            cartItem.setQuantity(cartItem.getQuantity() + requestDto.quantity());
+        }
+        return cartItemMapper.toDto(cartItemRepository.save(cartItem));
+    }
+
+    private CartItem createCartItemFromRequest(
```

Comment:
seems like this method does a mapper job

Reply:
In fact, this method uses a mapper and set the cart and book. I moved this functionality into a separate method in order to use it as an cart item creation in `orElseGet` if an item with such a book was not found in the database.